hudi-common/src/main/java/org/apache/hudi/common/table/read/ReusableKeyBasedRecordBuffer.java

```diff
@@ -0,0 +1,186 @@
+ * The implementation is used when point lookups are required for the same file group, such as the metadata table's files partition.
+ * As a result, the buffer also requires a set of valid keys to be provided to avoid returning all records in the processed logs when doing these point or small-batch lookups.
+ * @param <T> the engine specific record type
+ */
+public class ReusableKeyBasedRecordBuffer<T> extends FileGroupRecordBuffer<T> {
+  private final Set<String> validKeys;
+  private final Map<Serializable, BufferedRecord<T>> existingRecords;
+
+  ReusableKeyBasedRecordBuffer(HoodieReaderContext<T> readerContext, HoodieTableMetaClient hoodieTableMetaClient,
+                               RecordMergeMode recordMergeMode, PartialUpdateMode partialUpdateMode,
+                               TypedProperties props, Option<String> orderingFieldName,
+                               UpdateProcessor<T> updateProcessor, Map<Serializable, BufferedRecord<T>> records) {
+    this(readerContext, hoodieTableMetaClient, recordMergeMode, partialUpdateMode, props, orderingFieldName, updateProcessor, records, Collections.emptySet());
+  }
+
+  private ReusableKeyBasedRecordBuffer(HoodieReaderContext<T> readerContext, HoodieTableMetaClient hoodieTableMetaClient,
+                                       RecordMergeMode recordMergeMode, PartialUpdateMode partialUpdateMode,
+                                       TypedProperties props, Option<String> orderingFieldName,
+                                       UpdateProcessor<T> updateProcessor, Map<Serializable, BufferedRecord<T>> records, Set<String> validKeys) {
+    super(readerContext, hoodieTableMetaClient, recordMergeMode, partialUpdateMode, props, orderingFieldName, updateProcessor);
+    this.existingRecords = records;
+    this.validKeys = validKeys;
+  }
+
+  public ReusableKeyBasedRecordBuffer<T> withKeyPredicate(Option<Predicate> keyFilter) {
+    ValidationUtils.checkArgument(keyFilter.get() instanceof Predicates.In, () -> "Key filter should be of type Predicates.In, but found: " + keyFilter.get().getClass().getName());
+    List<Expression> children = ((Predicates.In) keyFilter.get()).getRightChildren();
+    Set<String> validKeys = children.stream().map(e -> (String) e.eval(null)).collect(Collectors.toSet());
+    return new ReusableKeyBasedRecordBuffer<>(readerContext, hoodieTableMetaClient, recordMergeMode, partialUpdateMode,
+        props, orderingFieldName, updateProcessor, existingRecords, validKeys);
+  }
+
+  @Override
+  protected ExternalSpillableMap<Serializable, BufferedRecord<T>> initializeRecordsMap(String spillableMapBasePath) {
+    return (ExternalSpillableMap<Serializable, BufferedRecord<T>>) existingRecords;
+  }
+
+  @Override
+  protected void initializeLogRecordIterator() {
+    logRecordIterator = new RemainingRecordIterator<>(validKeys, existingRecords);
```

Comment:
In the standard key based record buffer, we will iterate through the remaining keys to process records that are only present in the log files. This record buffer will filter those based on the key filter.